docs/oppen_algorithm.md

```diff
@@ -0,0 +1,179 @@
+| Begin     | Opens a group.
+| End       | Closes a group.
+| Break     | Potential line break (adds a new line if the text doesn't fit on the line).
+| LineBreak | Guaranteed line break.
+
+### Groups
+
+A group represents a sequence of items that shall be printed together.
+A consistent group propagates a new line to all the `BREAK` tokens in the group,
+while an inconsistent group does not propagate the new line.
+
+## The algorithm
+
+The algorithm can be divided into two parts: the `scanning` and the `printing`. As their name implies, the former adds new tokens into our buffers/stacks (expanding to the right) while the latter pushes onto our `print stack` the tokens from the bottom of our buffers (shrinking from the left).
+We scan until the current length of the line is greater or equal than the maximum desired width, then we will keep on pushing onto the `print stack` the tokens that we pop from our `tokens queue` while their correspondong `size buffer` value is greater than 0.
+
+### Scan stack
+
+The `scan stack` saves the locations of all the `BEGIN`, `END`, and `BREAK` tokens, allowing the algorithm to quickly access the last `BREAK` tokens and the groups they belong to.
+
+### Print stack
+The `print stack` handles the display of the tokens, taking into account the current indentation and the available space left on the line.
+It saves only the `BEGIN` tokens, the other tokens are only printed to the output.
+
+### Tokens queue
+
+The `tokens queue`'s saves the scanned tokens that are not pushed yet to the `print stack`.
+
+### Size buffer
+
+The `size buffer` stores at the `idx` position:
+  - The width of the token if it is a `STRING` token.
+  - The required width to print the previous tokens on the line until this point as a negative otherwise.
+
+Since there is no need to compute the exact size of the line if it is greater than the maximum desired width, we add `∞` in the `size buffer` instead.
+
+### How it works
+
+Here is a step by step execution of the algorithm using the example below.
+```
+[
+  BEGIN#1,
+  STRING#1 = 'Hello',
+  BREAK    = ', ',
+  STRING#2 = 'World!!!!!',
+
+  BEGIN#2,
+  LINE_BREAK,
+  STRING#3 = 'GoodBye, ',
+  STRING#4 = 'World!!!!!',
+  END#1,
+
+  END#2,
+
+  EOF,
+]
+```
+
+By processing the tokens one by one until `STRING#3`, we will get the following values in our data structures:
+```
++--------------+---------+----------+-------+----------+---------+-----------+----------+
+| tokens_queue | BEGIN#1 | STRING#1 | BREAK | STRING#2 | BEGIN#2 | LINEBREAK | STRING#3 |
++--------------+---------+----------+-------+----------+---------+-----------+----------+
+| size_buffer  |   -1    |    5     |  -6   |    10    |   -18   |    -18    |    9     |
++--------------+---------+----------+-------+----------+---------+-----------+----------+
+
++-------------+-------+-------+-------+-------+
+| scan_stack  |   0   |   2   |   4   |   5   |
++-------------+-------+-------+-------+-------+
+```
+
```

Comment:
I think it's a bad idea to:

1. Not indicate the desired width
2. Start with filled stacks and buffers. The size buffer is especially confusing.
3. Skip the print stack.